lightning/src/routing/router.rs

```diff
@@ -333,79 +429,199 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
-					if let Some(node_info) = node.announcement_info.as_ref() {
-						res.last_mut().unwrap().node_features = node_info.features.clone();
+		// At this point, targets are filled with the data from first and last hops communicated by the caller, and the payment receiver.
+		'path_construction: while let Some(RouteGraphNode { pubkey, lowest_fee_to_node, .. }) = targets.pop() {
+
+			// Since we're going dest-to-source, hitting our node as a target means that we should stop traversing the
+			// graph and arrange the path out of what we found.
+			if pubkey == *our_node_id {
+				let mut new_entry = dist.remove(&our_node_id).unwrap();
+				let mut res = vec!(new_entry.route_hop.clone());
+				let mut path_bottleneck_msat = None;
+				let mut cur_path_fees = Vec::new();
+				let mut smallest_channel_id = None;
+
+				loop {
+					if let Some(&(_, ref features)) = first_hop_targets.get(&res.last().unwrap().pubkey) {
+						res.last_mut().unwrap().node_features = features.to_context();
+					} else if let Some(node) = network.get_nodes().get(&res.last().unwrap().pubkey) {
+						if let Some(node_info) = node.announcement_info.as_ref() {
+							res.last_mut().unwrap().node_features = node_info.features.clone();
+						} else {
+							res.last_mut().unwrap().node_features = NodeFeatures::empty();
+						}
 					} else {
-						res.last_mut().unwrap().node_features = NodeFeatures::empty();
+						// We should be able to fill in features for everything except the last
+						// hop, if the last hop was provided via a BOLT 11 invoice (though we
+						// should be able to extend it further as BOLT 11 does have feature
+						// flags for the last hop node itself).
+						assert!(res.last().unwrap().pubkey == *target);
 					}
-				} else {
-					// We should be able to fill in features for everything except the last
-					// hop, if the last hop was provided via a BOLT 11 invoice (though we
-					// should be able to extend it further as BOLT 11 does have feature
-					// flags for the last hop node itself).
-					assert!(res.last().unwrap().pubkey == *target);
+
+					// For this path, select the channel with the lowest available amount to avoid hitting it
+					// in the next path finding iterations.
+					if let Some(current_hop_bottleneck_msat) = new_entry.available_msat {
+						if let Some(full_path_bottleneck_msat) = path_bottleneck_msat {
+							if current_hop_bottleneck_msat < full_path_bottleneck_msat {
+								path_bottleneck_msat = Some(current_hop_bottleneck_msat);
+								smallest_channel_id = Some(res.last().unwrap().short_channel_id);
+							}
+						}
+					}
+
+					if res.last().unwrap().pubkey == *target {
+						break;
+					}
+
+					new_entry = match dist.remove(&res.last().unwrap().pubkey) {
+						Some(hop) => hop,
+						None => return Err(LightningError{err: "Failed to find a non-fee-overflowing path to the given destination".to_owned(), action: ErrorAction::IgnoreError}),
```

Comment:
Is it possible we hit this and return early when calculating our Nth path, even though we have paths available that work?